lib/handlers/authorize-handler.js

```diff
@@ -193,10 +194,17 @@ AuthorizeHandler.prototype.getClient = function(request) {
         throw new InvalidClientError('Invalid client: missing client `redirectUri`');
       }
 
-      if (redirectUri && !client.redirectUris.includes(redirectUri)) {
-        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      if (redirectUri) {
```

Comment:
Why validating at this step and not below this line: https://github.com/node-oauth/node-oauth2-server/blob/98a9d197ef7c0d3f20cca65ffec59973f2089d66/lib/handlers/authorize-handler.js#L176

Reply:
I tried to make as minimal changes as possible to the existing code base. I also need the client which is a result of the promise beneath it.

Reply:
Sorry, I'm still a bit cloudy from a recent cold. Got it now. There is one thing though I would add to the tests, which are the [use-cases mentioned in the original issue](https://github.com/oauthjs/node-oauth2-server/pull/482#issuecomment-442682407):

> redirect_uri doesn't have to be exactly equal to any of the redirectUris. Conditions listed below are acceptable on some occasions.
> - same origin and same pathname but different query
> - same origin but different pathname
> - same domain but different protocol
> -  same top level domain

At least these should be covered by tests so we can make sure there is no follow-up issue with one of these use-cases, created by this new added code. What do you think?

Reply:
So the tests should implement `validateRedirectUri` and cover each use case?

Reply:
Hmm I am thinking now, since these use-cases are in any way user-specific implementations so I think we are fine here. Sometimes I am overthinking things :-D

Reply:
Okay I checked the code and I think at least some kind of implementation should be checked so we make sure the function arguments and return values are considered correctly, since currently it's only a stub, which is fine for unit tests but not sufficient in integration testing, what do you think?

Reply:
I agree. I updated the unit test.